gh-119993 ignore `NotADirectoryError` in `Path.unlink()` if `missing_ok` is `True`

MusicalNinjaDad opened the pull request:
Add a short note on when Path.unlink() raises `NotADirectoryError`

This can occur "in the wild", for example, after calling `shutil.move(src, dst)` where `dst` is a non-existent directory. Debugging can cost a lot of time as the cause can be non-intuitive, potentially not nearby in the code and the docs don't mention the exception as being raised in any circumstances.
(This just happened to me after encountering pypa/cibuildwheel#1850)

Fixes: #119993 

<!-- gh-issue-number: gh-119993 -->
* Issue: gh-119993
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--120049.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

MusicalNinjaDad committed:
Add a comment on when `NotADirectoryError`is raised by Path.unlink()

MusicalNinjaDad committed:
don't prefer any specific OSError type in Path.unlink documentation

barneygale commented on Doc/library/pathlib.rst:
```diff
@@ -1624,8 +1624,7 @@ example because the path doesn't exist).
 
-   If *missing_ok* is false (the default), :exc:`FileNotFoundError` is
-   raised if the path does not exist.
+   This method propagates any :exc:`OSError` encountered during removal.
```
```suggestion
   If *missing_ok* is false (the default), this method propagates any 
   :exc:`OSError` encountered during removal.
```

MusicalNinjaDad replied:
Hi @barneygale - is that technically accurate? My understanding is that any `OSError` is propagated regardless of `missing_ok`, which suppresses any `FileNotFoundError`.
I understand the interest to specifically reference the `FileNotFoundError` in the context of `missing_ok` ... how about this new suggestion?

MusicalNinjaDad replied:
```
   This method propagates any :exc:`OSError` encountered during removal. 
    For example: If *missing_ok* is false (the default), :exc:`FileNotFoundError` is 
     raised if the path does not exist. 
 
 
     If *missing_ok* is true, :exc:`FileNotFoundError` exceptions will be ...
```

barneygale replied:
`FileNotFoundError` is a subclass of `OSError`! Maybe we should make that clearer:

```suggestion
   If *missing_ok* is false (the default), this method propagates any 
   :exc:`OSError` from the operating system, including :exc:`FileNotFoundError`.
```

MusicalNinjaDad replied:
Thanks @barneygale, I've updated with that text and slightly extended the following paragraph:

```rst
   If *missing_ok* is false (the default), this method propagates any 
   :exc:`OSError` from the operating system, including :exc:`FileNotFoundError`.

   If *missing_ok* is true, :exc:`FileNotFoundError` exceptions will be
   ignored (same behavior as the POSIX ``rm -f`` command), any other
   :exc:`OSError` which is encountered will continue to be propogated.
```

MusicalNinjaDad committed:
maintain the original clarity that FileNotFound will be raised if missing_ok is false

MusicalNinjaDad committed:
clarify that FileNotFoundError is a subclass of OSError

MusicalNinjaDad committed:
fix whitespace

MusicalNinjaDad committed:
ignore NotADirectoryError on missing_ok = True

Semantically equivalent to FileNotFoundError - the file to be unlinked does not exist

MusicalNinjaDad committed:
Merge branch 'main' into MusicalNinjaDad/issue119993

MusicalNinjaDad committed:
Windows doesn't raise NotADirectoryError

barneygale commented on Doc/library/pathlib.rst:
```diff
@@ -1630,15 +1630,21 @@ Copying, renaming and deleting
-   If *missing_ok* is false (the default), :exc:`FileNotFoundError` is
-   raised if the path does not exist.
+   If *missing_ok* is false (the default), this method propagates any
+   :exc:`OSError` from the operating system, including :exc:`FileNotFoundError`.
```
Let's leave the details for the next paragraph.

```suggestion
   If *missing_ok* is false (the default), this method propagates any
   :exc:`OSError` from the operating system.
```

barneygale commented on Doc/library/pathlib.rst:
```diff
@@ -1630,15 +1630,21 @@ Copying, renaming and deleting
+   command and any :exc:`FileNotFoundError` or :exc:`NotADirectoryError`
+   exceptions will be ignored. This means that the file does not exist after
+   execution, but cannot guarantee that the file did exist before. Any other
+   :exc:`OSError` which is encountered will continue to be propogated.
```
I don't think we need to elaborate so much!

```suggestion
   If *missing_ok* is true, :exc:`FileNotFoundError` and 
   :exc:`NotADirectoryError` exceptions will be ignored. This behavior is
   similar to the POSIX ``rm -f`` command.
```

barneygale commented on Doc/library/pathlib.rst:
```diff
@@ -1630,15 +1630,21 @@ Copying, renaming and deleting
 
+   .. versionchanged:: 3.??
+      The *missing_ok* parameter will also ignore :exc:`NotADirectoryError`
+
```
```suggestion
   .. versionchanged:: 3.14
      Suppresses :exc:`NotADirectoryError` exceptions when *missing_ok* is
      true.
```

barneygale commented on Lib/test/test_pathlib/test_pathlib.py:
```diff
@@ -778,6 +778,35 @@ def test_group_no_follow_symlinks(self):
+        self.assertFileNotFound(p.unlink)
+        p.unlink(missing_ok=True)
+
+    def test_unlink_missing_ok_intermediate_file(self):
```
This should be added to `test_pathlib_abc.py` alongside the other `test_unlink_*` tests there. Those tests are inherited into this class.

barneygale commented on Lib/test/test_pathlib/test_pathlib.py:
```diff
@@ -778,6 +778,35 @@ def test_group_no_follow_symlinks(self):
+        p.touch()
+        p = p / 'fileBBB'
+        if sys.platform.startswith("win"):
+            self.assertFileNotFound(p.unlink)
```
Add a `@needs_posix` decorator to the test and don't bother testing Windows IMO :]

MusicalNinjaDad replied:
Does this hurt to have it in? Isn't it generally better to have the extra test coverage, or are we really struggling to keep execution times down?

MusicalNinjaDad committed:
make the documentation more succinct
implements comments from @barneygale

MusicalNinjaDad committed:
also run `test_unlink_missing_ok_intermediate_file` against `DummyPath`

MusicalNinjaDad committed:
DummyPath.unlink raises NotADirectoryError

MusicalNinjaDad committed:
Only test on posix

MusicalNinjaDad committed:
Merge remote-tracking branch 'origin/main' into MusicalNinjaDad/issue119993

Keep origin/main version of test_pathlib_abc.py - need to reimplement changes from MusicalNinjaDad/issue119993 in context of significant changes to class structure
